cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs

```diff
@@ -0,0 +1,2458 @@
+        public static Guid AddSolutionComponent(this CDSWebApiService svc,
+            Guid componentId,
+            SolutionComponentType componentType,
+            string solutionUniqueName,
+            bool addRequiredComponents,
+            bool? doNotIncludeSubComponents = null,
+            string includedComponentsSettingsValues = null)
+        {
+            Guid solutionComponentId;
+
+            if (!string.IsNullOrEmpty(solutionUniqueName))
+            {
+                //Add the attribute as part of a solution
+                JObject addSolutionComponentParameters = new JObject
+                {
+                    ["ComponentId"] = componentId,
+                    ["ComponentType"] = (int)componentType,
+                    ["SolutionUniqueName"] = solutionUniqueName,
+                    ["AddRequiredComponents"] = addRequiredComponents
+                };
+                if (doNotIncludeSubComponents != null)
+                {
+                    addSolutionComponentParameters["DoNotIncludeSubcomponents"] = doNotIncludeSubComponents;
+                }
+                if (includedComponentsSettingsValues != null)
+                {
+                    addSolutionComponentParameters["IncludedComponentSettingsValues"] = includedComponentsSettingsValues;
+                }
+
+                try
+                {
+                    JObject addSolutionComponentResponse = new JObject();
+
+                    addSolutionComponentResponse = svc.Post("AddSolutionComponent", addSolutionComponentParameters, STRONGCONSISTENCYHEADER);
+
+                    solutionComponentId = new Guid(addSolutionComponentResponse.GetValue("id").ToString());
+                }
+                catch (Exception ex)
+                {
+                    throw new Exception($"There was a problem adding the {componentType} as a component to the {solutionUniqueName} solution:{ex.Message}");
+                }
+            }
+            else
+            {
+                throw new Exception("AddSolutionComponent requires the solutionUniqueName parameter value.");
+            }
+            return solutionComponentId;
+        }
+
+        /// <summary>
+        /// Checks whether the specified entity can be the primary entity (one) in a one-to-many relationship.
+        /// </summary>
+        /// <param name="svc">An instance of the CDSWebApiService to extend.</param>
+        /// <param name="entityName">The logical name of the entity to check.</param>
+        /// <returns>Whether the specified entity can be the primary entity (one) in a one-to-many relationship.</returns>
+        public static bool CanBeReferenced(this CDSWebApiService svc, string entityName)
+        {
+            try
+            {
+                JObject body = new JObject()
+                {
+                    ["EntityName"] = entityName
+                };
+                return (bool)svc.Post("CanBeReferenced", body)["CanBeReferenced"];
+            }
```

Comment:
General observation that is replicated throughout this code file. You have try/catch blocks in which you create a new exception with the caught exception as the inner exception and the parent exception message contains the inner exception message. Or you create an exception that tells the end user what method the exception occurred in. Neither of these are needed as you will be able to find this info when using a debugger.

Perhaps you like having such info printed out in the console without having to use a debugger? The problem with this approach is that 1) our dev customers know and expect to use a debugger, and 2) if they want to copy and use the code in their own projects (which they do) they will have to strip this code out.

Reply:
I'd be happy to discuss alternative patterns. The error handling I've done is focused on identifying the source of the error while debugging the console application sample. This is somewhere in-between the raw-low level samples and an SDK.

If you have a different pattern that you would recommend, please submit a PR and I can review it.